File: PokeLangPython/Lexer/MathOperator.py
from Token import Token

class MathOperator:
    def evaluate(self, code, linha, column):
        code_iter = iter(code)
        for _ in code:
            current_char = next(code_iter, None)
            if current_char == '-':
                return Token("SUB", "-", linha, column)
            elif current_char == '*':
                return Token("MUL", "*", linha, column)
            elif current_char == '/':
                return Token("DIV", "/", linha, column)
            elif current_char == '(':
                return Token("LPAREN", "(", linha, column)
            elif current_char == ')':
                return Token("RPAREN", ")", linha, column)
            elif current_char == '{':
                return Token("LCHAVE", "{", linha, column)
            elif current_char == '[':
                return Token("LCOLCHETE", "[", linha, column)
            elif current_char == ']':
                return Token("rCOLCHETE", "]", linha, column)
            elif current_char == '}':
                return Token("RCHAVE", "}", linha, column)
            elif current_char == ';':
                return Token("FIM", ";", linha, column)
            elif current_char == '>':
                return Token("MAIORQ", ">", linha, column)
            elif current_char == '<':
                return Token("MENORQ", "<", linha, column)
            elif current_char == '&':
                next_char = next(code_iter, None)
                if next_char == '&':
                    return Token("AND", "&&", linha, column)
            elif current_char == '|':
                next_char = next(code_iter, None)
                if next_char == '|':
                    return Token("AND", "||", linha, column)
            elif current_char == '=':
                next_char = next(code_iter, None)
                if next_char == '=':
                    return Token("IGUAL", "==", linha, column)
                else:
                    return Token("ATRIBUICAO", "=", linha, column)
            elif current_char == '!':
                next_char = next(code_iter, None)
                if next_char == '=':
                    return Token("DIFERENTE", "!=", linha, column)
            elif current_char == '+':
                next_char = next(code_iter, None)
                if next_char == '+':
                    return Token("INCREMENTO_SOMA", "++", linha, column)
                else:
                    return Token("SOMA", "+", linha, column)
        return None

    def skip_white_space(self, code):  # Adiciona esta função
        while code.current() in (' ', '\n', '\r', '\t'):
            code.next()

File: PokeLangJava/src/AFD/MathOperator.java
package AFD;

import java.text.CharacterIterator;
import Token.Token;

public class MathOperator extends AFD {
    @Override
    public Token evaluate(CharacterIterator code) {
        switch (code.current()) {
            case '-':
                code.next();
                return new Token("SUB", "-");
            case '*':
                code.next();
                return new Token("MUL", "*");
            case '/':
                code.next();
                return new Token("DIV", "/");
            case '(':
                code.next();
                return new Token("LPAREN", "(");
            case ')':
                code.next();
                return new Token("RPAREN", ")");
            case '{':
                code.next();
                return new Token("LCHAVE", "{");
            case '[':
                code.next();
                return new Token("LCOLCHETE", "[");
            case ']':
                code.next();
                return new Token("rCOLCHETE", "]");
            case '}':
                code.next();
                return new Token("RCHAVE", "}");
            case ';':
                code.next();
                return new Token("FIM", ";");
            case '>':
                code.next();
                return new Token("MAIORQ", ">");
            case '<':
                code.next();
                return new Token("MENORQ", "<");
            case '&':
                code.next();
                if (code.current() == '&') {
                    code.current();
                    return new Token("AND", "&&");
                }
            case '|':
                code.next();
                if (code.current() == '|') {
                    code.next();
                    return new Token("AND", "||");
                }
            case '=':
                code.next();
                if (code.current() == '=') {
                    code.next();
                    return new Token("IGUAL", "==");
                } else {
                    return new Token("ATRIBUICAO", "=");
                }
            case '!':
                code.next();
                if (code.current() == '=') {
                    code.next();
                    return new Token("DIFERENTE", "!=");
                }
            case '+':
                code.next();
                if (code.current() == '+') {
                    code.next();
                    return new Token("INCREMENTO_SOMA", "++");
                } else {
                    return new Token("SOMA", "+");
                }
            default:
                return null;
        }
    }
}
